pkg/kwok/server/debugging_exec.go

```diff
@@ -21,23 +21,108 @@ import (
 	"fmt"
 	"io"
 	"net/http"
-	"time"
 
 	"github.com/emicklei/go-restful/v3"
 	"k8s.io/apimachinery/pkg/types"
 	clientremotecommand "k8s.io/client-go/tools/remotecommand"
 
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
 	"sigs.k8s.io/kwok/pkg/kwok/server/remotecommand"
+	"sigs.k8s.io/kwok/pkg/utils/exec"
+	"sigs.k8s.io/kwok/pkg/utils/slices"
 )
 
 // ExecInContainer executes a command in a container.
-func (s *Server) ExecInContainer(ctx context.Context, podName, podNamespace string, uid types.UID, container string, cmd []string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan clientremotecommand.TerminalSize, timeout time.Duration) error {
-	// TODO: Configure and implement the exec streamer
-	msg := fmt.Sprintf("TODO: ExecInContainer(%q, %q, %q)", podName+"/"+podNamespace, container, cmd)
-	_, _ = out.Write([]byte(msg))
+func (s *Server) ExecInContainer(ctx context.Context, podName, podNamespace string, uid types.UID, container string, cmd []string, in io.Reader, out, errOut io.WriteCloser, tty bool, resize <-chan clientremotecommand.TerminalSize) error {
+	execTarget, err := s.getExecTarget(podName, podNamespace, container)
+	if err != nil {
+		return err
+	}
+
+	// Currently only support local exec.
+	if execTarget.Local == nil {
+		return fmt.Errorf("not set local exec")
+	}
+
+	// Set the environment variables.
+	if len(execTarget.Local.Envs) != 0 {
+		envs := slices.Map(execTarget.Local.Envs, func(env internalversion.EnvVar) string {
+			return fmt.Sprintf("%s=%s", env.Name, env.Value)
+		})
+		ctx = exec.WithEnv(ctx, envs)
+	}
+
+	// Set the working directory.
+	if execTarget.Local.WorkDir != "" {
+		ctx = exec.WithDir(ctx, execTarget.Local.WorkDir)
+	}
+
+	// Set cancel context.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if tty {
+		return s.execInContainerWithTTY(ctx, cmd, in, out, errOut, resize)
+	}
+
+	return s.execInContainer(ctx, cmd, in, out, errOut)
+}
+
+func (s *Server) execInContainer(ctx context.Context, cmd []string, in io.Reader, out, errOut io.WriteCloser) error {
+	// Set the stream as the stdin/stdout/stderr.
+	ctx = exec.WithIOStreams(ctx, exec.IOStreams{
+		In:     in,
+		Out:    out,
+		ErrOut: errOut,
+	})
+
+	// Execute the command.
+	err := exec.Exec(ctx, cmd[0], cmd[1:]...)
+	if err != nil {
+		return err
+	}
 	return nil
 }
 
+func (s *Server) getExecTarget(podName, podNamespace string, containerName string) (*internalversion.ExecTarget, error) {
+	pf, has := slices.Find(s.config.Execs, func(pf *internalversion.Exec) bool {
+		return pf.Name == podName && pf.Namespace == podNamespace
+	})
+	if has {
+		exec, found := findContainerInExecs(containerName, pf.Spec.Execs)
+		if found {
+			return exec, nil
+		}
+	} else {
```

Comment:
Question: If we find exec that is misconfigured (for example, pod name and pod namespace are matching, but there are no containers named `containerName`, should we fall back to the cluster execs or return not found?
This is also important for the logging PR :smile: 

If we should return an error, I'd suggest return early and remove else
Then the error message can also specify where did we stop execution so it is more understandable to users where to look for.
For example, saying that pod matching exec spec does not have a container with container name can be a signal that the `Exec` configuration is wrong

Reply:
makes sense 😄